src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+
+    console.groupCollapsed(`%cblazor%c Loaded ${toDataSizeString(totalResponseBytes)} resources${linkerDisabledWarning}`, 'background: purple; color: white; padding: 1px 3px; border-radius: 3px;', 'font-weight: bold;', 'font-weight: normal;');
+
+    if (cacheLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(cacheResponseBytes)} resources from cache`);
+      console.table(this.cacheLoads);
+      console.groupEnd();
+    }
+
+    if (networkLoadsEntries.length) {
+      console.groupCollapsed(`Loaded ${toDataSizeString(networkResponseBytes)} resources from network`);
+      console.table(this.networkLoads);
+      console.groupEnd();
+    }
+
+    console.groupEnd();
+  }
+
+  async purgeUnusedCacheEntriesAsync() {
+    // We want to keep the cache small because, even though the browser will evict entries if it
+    // gets too big, we don't want to be considered problematic by the end user viewing storage stats
+    const cachedRequests = await this.cache.keys();
+    const deletionPromises = cachedRequests.map(async cachedRequest => {
+      if (!(cachedRequest.url in this.usedCacheKeys)) {
+        await this.cache.delete(cachedRequest);
+      }
+    });
+
+    return Promise.all(deletionPromises);
+  }
+
+  private async tryValidateAndCacheAsync(name: string, cacheKey: string, expectedContentHash: string, networkResponse: Response): Promise<ArrayBuffer> {
+    const responseBuffer = await networkResponse.arrayBuffer();
+
+    // Now is an ideal moment to capture the performance stats for the request, since it
+    // only just completed and is most likely to still be in the buffer. However this is
+    // only done on a 'best effort' basis. Even if we do receive an entry, some of its
+    // properties may be blanked out if it was a CORS request.
+    const performanceEntry = getPerformanceEntry(networkResponse.url);
+    const responseBytes = (performanceEntry && performanceEntry.encodedBodySize) || undefined;
+    this.networkLoads[name] = { responseBytes };
+
+    // crypto.subtle is only enabled on localhost or HTTPS origins
+    // We only write to the cache if we can validate the content hashes
+    if (typeof crypto !== 'undefined' && !!crypto.subtle) {
+      await assertContentHashMatchesAsync(name, responseBuffer, expectedContentHash);
+
+      // Build a custom response object so we can track extra data such as responseBytes
+      // We can't rely on the server sending content-length (ASP.NET Core doesn't by default)
+      await this.cache.put(cacheKey, new Response(responseBuffer, {
+        headers: {
+          'content-type': networkResponse.headers.get('content-type') || '',
+          'content-length': (responseBytes || networkResponse.headers.get('content-length') || '').toString()
+        }
+      }));
+    }
+
+    return responseBuffer;
```

Comment:
Instead of validating the contents of the cache when we download bits from the network, can we validate them when we load them from the cache? Or as an asynchronous task that happens after we've downloaded and cached the resources? If we detect this situation, we can simply crash the app at that point, isn't that correct?

It is very unlikely that you end up in a situation where the manifest doesn't reflect the contents of the server or that you can't trust the manifest (it's your app, you are serving it).

Worst case scenario, your app is in the middle of a deployment and you got the wrong contents, the likely outcome is that your app will crash and the user will have to reload (this problem is orthogonal to what we are trying here).

If you want to do this validation, then do it over the cached data the first time you load from cache and "repair" purge the entry from the dll once you detect that scenario.

Reply:
> Instead of validating the contents of the cache when we download bits from the network, can we validate them when we load them from the cache?

That seems strictly worse to me. We're then doing the hash check on every load instead of only when the data was first acquired. We still have to do it on the first load when it comes from the network too (see below).

> It is very unlikely that you end up in a situation where the manifest doesn't reflect the contents of the server or that you can't trust the manifest (it's your app, you are serving it).

It will certainly happen sometimes. It's always possible during deployments that a user fetches `blazor.boot.json` from the old version of the site and some assemblies from the new version.

> Worst case scenario, your app is in the middle of a deployment and you got the wrong contents, the likely outcome is that your app will crash and the user will have to reload (this problem is orthogonal to what we are trying here)

Doesn't seem like an orthogonal problem to me. The scenario this code deals with is acquiring the resources and verifying they are the ones we want. I don't see any benefit in letting the app start up with the wrong resources, performing whatever undefined behavior this leads to for some period, and then hoping we can shut it down later.

> If you want to do this validation, then do it over the cached data the first time you load from cache and "repair" purge the entry from the dll once you detect that scenario.

I definitely think we should validate the data the first time it comes from the network, for the reasons mentioned above. And then having done that and cached it under a key derived from that hash, there's no purpose in revalidating it again later.

Reply:
Has this conversation been rendered obsolete by our plan to use `{integrity: ....}`. I imagine we'll break at that point.

Reply:
I'm pretty confident this will get simpler when using `integrity`. Exactly how it will look after, I'm not certain yet. Shouldn't be a big job to implement.

Reply:
Done that. It is simpler now, and follows the natural pattern for integrity validation built into the `fetch` API.